R/create_demographic_lookup.R

```diff
@@ -0,0 +1,422 @@
+        recid, deathdiag1, deathdiag2, deathdiag3, deathdiag4, deathdiag5,
+        deathdiag6, deathdiag7, deathdiag8, deathdiag9, deathdiag10,
+        deathdiag11
+      )
+    ) %>%
+    assign_substance_cohort() %>%
+    # Aggregate to CHI level
+    dplyr::group_by(.data$chi) %>%
+    dplyr::summarise(dplyr::across(c(
+      "mh",
+      "frail",
+      "maternity",
+      "high_cc",
+      "medium_cc",
+      "low_cc",
+      "comm_living",
+      "adult_major",
+      "child_major",
+      "end_of_life",
+      "substance"
+    ), any)) %>%
+    dplyr::ungroup() %>%
+    # Assign demographic_cohort based on hierarchy of each cohort
+    dplyr::mutate(demographic_cohort = dplyr::case_when(
+      end_of_life ~ "End of Life",
+      frail ~ "Frailty",
+      high_cc ~ "High Complex Conditions",
+      maternity ~ "Maternity and Healthy Newborns",
+      mh ~ "Mental Health",
+      substance ~ "Substance Misuse",
+      medium_cc ~ "Medium Complex Conditions",
+      low_cc ~ "Low Complex Conditions",
+      child_major ~ "Child Major Conditions",
+      adult_major ~ "Adult Major Conditions",
+      comm_living ~ "Assisted Living in the Community",
+      TRUE ~ "Healthy and Low User"
+    )) %>%
+    # Reorder variables
+    dplyr::relocate(.data$demographic_cohort, .after = chi)
+
+  # Write to disk
+  if (write_to_disk == TRUE) {
+    write_rds(demo_lookup,
+      path = glue::glue("{get_slf_dir()}/Cohorts/Demographic_Cohorts_{year}.rds")
+    )
+  }
+
+  return(demo_lookup)
+}
+
+#' Assign Mental Health cohort
+#' @description A record is considered to be in the MH cohort if the recid is 04B. Also,
+#' if the recid is one of 01B, GLS, 50B, 02B or AE2 \strong{and} any of the diagnosis codes start with
+#' F2, F3, F067, F070, F072, F078, or F079
+#'
+#' @param recid A vector of record IDs
+#' @param diag1 A vector of Diagnosis Code 1
+#' @param diag2 A vector of Diagnosis Code 2
+#' @param diag3 A vector of Diagnosis Code 3
+#' @param diag4 A vector of Diagnosis Code 4
+#' @param diag5 A vector of Diagnosis Code 5
+#' @param diag6 A vector of Diagnosis Code 6
+#'
+#' @return A boolean vector indicating whether a given record is in the Mental Health cohort
+#' @family Demographic and Service Use Cohort functions
+assign_mh_cohort <- function(recid, diag1, diag2, diag3, diag4, diag5, diag6) {
+  mh <-
+    # FOR FUTURE: when variable MentalHealthProblemsClientGroup exists and is "Y", mh_cohort = TRUE
+    recid == "04B" |
+      (recid %in% c("01B", "GLS", "50B", "02B", "AE2") &
+        (rowSums(dplyr::across(
+          c("diag1", "diag2", "diag3", "diag4", "diag5", "diag6"),
+          ~ stringr::str_sub(.x, 1, 2) %in%
+            c("F2", "F3")
+        )) > 0 |
+          rowSums(dplyr::across(
+            c("diag1", "diag2", "diag3", "diag4", "diag5", "diag6"),
+            ~ stringr::str_sub(.x, 1, 4) %in%
+              c("F067", "F070", "F072", "F078", "F079")
+          )) > 0))
+  return(mh)
+}
```

Comment:
This whole thing was going quite slowly. This was the best I could do to speed it up, which can be applied to most / all the other functions. It's about a 30% speedup, I think its doing the `rowSums` + `across` bit once which makes the most difference.

```suggestion
assign_mh_cohort <- function(recid, diag1, diag2, diag3, diag4, diag5, diag6) {
  mh <-
    # FOR FUTURE: when variable MentalHealthProblemsClientGroup exists and is "Y", mh_cohort = TRUE
    dplyr::case_when(
    recid == "04B" ~ TRUE,
    recid %in% c("01B", "GLS", "50B", "02B", "AE2") &
      (rowSums(dplyr::across(
        c("diag1", "diag2", "diag3", "diag4", "diag5", "diag6"),
        ~ stringr::str_starts(.x, paste("F2", "F3", "F067", "F070", "F072", "F078", "F079", sep = "|")))) > 0) ~ TRUE,
      TRUE ~ FALSE
  )
  
  return(mh)
}
```